ecl/hql/hqlfilter.cpp

```diff
@@ -371,6 +371,47 @@ static IHqlExpression * createExpandedRecord(IHqlExpression * expr)
 }
 
 
+IHqlExpression * castToFieldAndBack(IHqlExpression * left, IHqlExpression * right)
+{
+    node_operator op = left->getOperator();
+    switch (op)
+    {
+    case no_cast:
+    case no_implicitcast:
+        {
+            IHqlExpression * uncast = left->queryChild(0);
+            ITypeInfo * castType = right->queryType();
+            ITypeInfo * uncastType = uncast->queryType();
+
+            OwnedHqlExpr castRight = ensureExprType(right, uncastType);
+            OwnedHqlExpr base = castToFieldAndBack(uncast, castRight);
+            //If this cast doesn't lose any information and child didn't change then don't bother
+            //casting back and forwards.
+            if ((base == castRight) && !castLosesInformation(uncastType, castType))
+                return LINK(right);
+            return ensureExprType(base, castType);
+        }
+
+    case no_select:
+        {
+            ITypeInfo * leftType = left->queryType();
+            ITypeInfo * rightType = right->queryType();
+            if (leftType == rightType || !castLosesInformation(leftType, rightType))
+                return LINK(right);
+            OwnedHqlExpr castToField = ensureExprType(right, leftType);
+            return ensureExprType(castToField, rightType);
+        }
+    case no_substring:
+    case no_add:
+    case no_sub:
+        return castToFieldAndBack(left->queryChild(0), right);
+    default:
+        UNIMPLEMENTED;
```

Comment:
Is this something that needs to be implemented (and if so when) or should it be a throwUnexpected() ?

Reply:
(I realize this is not new code, so may choose to ignore)